src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -131,28 +132,231 @@ namespace xocl {
+
+        /* no io submission error or more pending i/o requests */
+        if (!cbErrCnt || (cbSubmitCnt != cbPollCnt))
+            return 0;
+
+        int num_evt = (cbErrCnt <= nr_comps) ? cbErrCnt : nr_comps;
+        cbErrCnt -= num_evt;
+
+        int i = 0;
+        auto end = reqList.end();
+        for (auto it = reqList.begin(); it != end && i < num_evt; ++it, i++) {
+            comps[i].nbytes = 0;
+            comps[i].err_code = cbErrCode;
+            comps[i].priv_data = (void *)(*it).priv_data;
+        }
+
+        release_request(i);
+
+        if (!cbErrCnt)
+            cbErrCode = 0;
+
+        return num_evt;
+    }
+
+    /* prepare io submission structures */
+    void prepare_io(struct iocb *cb, struct iovec *iov,
+                    struct xocl_qdma_req_header *header,
+		    unsigned long buf_va, unsigned long buf_len,
+                    uint64_t priv_data)
+    {
+        iov[0].iov_base = header;
+        iov[0].iov_len = sizeof(struct xocl_qdma_req_header);
+        iov[1].iov_base = (void *)buf_va;
+        iov[1].iov_len = buf_len;
+
+        if (cb) {
+            memset(cb, 0, sizeof(struct iocb));
+            cb->aio_fildes = (int)qhndl;
+            cb->aio_lio_opcode = h2c ? IOCB_CMD_PWRITEV : IOCB_CMD_PREADV;
+            cb->aio_buf = (uint64_t)iov;
+            cb->aio_offset = 0;
+            cb->aio_nbytes = 2;
+            cb->aio_data = priv_data;
+        }
+    }
+
+    /* submit all of the queued i/o, calling function should hold the lock */
+    int queue_flush_aio_request(void)
+    {
+        /* if there is submission error, wait till all requests are drained */
+        if (cbErrCnt)
+            return 0;
+
+        /* submit all queued requests */
+        unsigned int cb_max = bufCnt;
+        unsigned int cb_cnt = 0;
+        std::vector <struct iocb *> cbpp;
+
+        auto end = reqList.end();
+        for (auto it = reqList.begin(); it != end && cb_cnt < cb_max;
+             ++it, cb_cnt++) {
+            auto qio = (*it);
+
+            qio.header.flags = qio.flags;
+            prepare_io(&qio.cb, qio.iov, &qio.header, qio.buf_va, qio.len,
+			qio.priv_data);
+            cbpp.push_back(&qio.cb);
+        }
+
+        int submitted = io_submit(qAioCtx, cb_cnt, cbpp.data());
+        if (submitted < 0) {
+            if (submitted != -EAGAIN) {
+		/* something went wrong, flush all of the queued request with
+                 * this error */
+                cbErrCnt = reqList.size();
+                cbErrCode= submitted;
+            }
+            return submitted;
+        }
+
+        if (submitted > 0) {
+            release_request(submitted);
+            cbSubmitCnt += submitted;
+        }
+        return 0;
+    }
+
+    /* worker thread for aio batching */
+    void queue_aio_worker(void)
+    {
+        byteCnt = bufCnt = 0;
+        cbSubmitCnt = cbPollCnt = cbErrCnt = 0;
+
+        std::mutex mtx;
+        std::unique_lock<std::mutex> lck(mtx);
```

Comment:
Are these really needed.  Should the loop just acquire reqLock and wait on that lck?  I doubt the 'if .. else ...' with cv.wait has any significant time that allows reqLock to be locked by other thread before it is immediately locked at line 296 in the loop.  In short, I would have expected:
```
do {
  std::unique_lock<std::mutex> lck(reqLock);
  if ...
  else ..
  ...
}
```
with no additional locking of reqLock.

Reply:
I don't want to hold the lock while waiting for the timer to pop. A user could submit an i/o during this time and it can be added to the reqList.